src/operator/tensor/matrix_op-inl.h

```diff
@@ -476,6 +477,405 @@ void DotBackward_(const nnvm::NodeAttrs& attrs,
+                s, data_out.Size(), data_out.dptr<DType>(), row_idx_out.dptr<RType>(),
+                data_l.dptr<DType>(), indptr_l.dptr<IType>(), col_idx_l.dptr<CType>(),
+                data_r.dptr<DType>(), data_out.shape_[1]); 
+          }
+        });
+      });
+    });
+  });
+}
+
+template<typename xpu>
+void DotForwardEx(const nnvm::NodeAttrs& attrs,
+                  const OpContext& ctx,
+                  const std::vector<NDArray>& inputs,
+                  const std::vector<OpReqType>& req,
+                  const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const DotParam& param = nnvm::get<DotParam>(attrs.parsed);
+  CHECK(!param.transpose_b) << "tranposing rhs of the op dot is not supported";
+
+  NDArray ret = outputs[0];  // get rid of the const qualifier
+  if (inputs[0].storage_type() == kCSRStorage
+      && inputs[1].storage_type() == kDefaultStorage
+      && outputs[0].storage_type() == kDefaultStorage) {
+    DotCsrDnsDnsImpl<xpu>(ctx, inputs[0], inputs[1], req[0], param.transpose_a, &ret);
+  } else if (inputs[0].storage_type() == kCSRStorage
+      && inputs[1].storage_type() == kDefaultStorage
+      && outputs[0].storage_type() == kRowSparseStorage) {
+    DotCsrDnsRspImpl<xpu>(ctx, inputs[0], inputs[1], req[0], param.transpose_a, &ret);
+  } else {
+    // TODO(junwu): add fallback mechanism
+    LOG(FATAL) << "Not supported";
+  }
+}
+
+/*!
+ * \brief Backward of
+ * 1. out = dot(csr, dns)
+ * grad(csr) = dot(grad(out), dns.T())
+ * grad(dns) = dot(csr.T(), grad(out))
+ *
+ * 2. out = dot(csr.T(), dns)
+ * grad(csr) = dot(dns, grad(out).T())
+ * grad(dns) = dot(csr, grad(out))
+ *
+ * Assume the gradient of the op's output is a dense matrix.
+ * This function must be called by DotBackwardEx.
+ */
+template<typename xpu>
+void DotBackwardCsrDnsRsp(const nnvm::NodeAttrs& attrs,
+                          const OpContext& ctx,
+                          const std::vector<NDArray>& inputs,
+                          const std::vector<OpReqType>& req,
+                          const std::vector<NDArray>& outputs) {
+  const DotParam& param = nnvm::get<DotParam>(attrs.parsed);
+  NDArray ret = outputs[1];
+  DotCsrDnsRspImpl<xpu>(ctx, inputs[1], inputs[0], req[1], !param.transpose_a, &ret);
+}
+
+template<typename xpu>
+void DotBackwardEx(const nnvm::NodeAttrs& attrs,
+                   const OpContext& ctx,
+                   const std::vector<NDArray>& inputs,
+                   const std::vector<OpReqType>& req,
+                   const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 2U);
+  CHECK_EQ(req.size(), 2U);
+  CHECK_EQ(kNullOp, req[0])
+    << "sparse dot does not support computing the gradient of the csr/lhs";
+  CHECK_NE(req[1], kWriteInplace) << "DotBackwardCsrDnsRsp does not support WriteInplace";
```

Comment:
Current for sparse ndarray we only have kWriteTo.. This is probably not ideal. We should revisit it in the future

https://github.com/eric-haibin-lin/mxnet/blob/master/src/executor/inplace_addto_detect_pass.cc#L48
https://github.com/eric-haibin-lin/mxnet/blob/master/src/executor/graph_executor.cc#L675